CIP-MultiScriptHashes/README.md

```diff
@@ -0,0 +1,95 @@
+---
+CIP: "???"
+Title: Multi-Script Hashes
+Authors: Mario Blažević <mario@mlabs.city>
+Discussions-To: TBD
+Comments-Summary: TBD
+Comments-URI: TBD
+Status: Draft
+Type: Standards Track  
+Created: <date created on, in ISO 8601 (yyyy-mm-dd) format>  
+License: CC-BY-4.0  
+License-Code: Apache-2.0  
+Post-History: [2022-06-09](https://forum.cardano.org/t/cip-draft-multi-script-hashes/102845)
+---
+
+## Abstract
+
+This is a (draft) CIP to introduce a simple way to statically identify a collection of scripts that depend on each
+other, so that each script can detect if a transaction tries to substitute an outside script address for an expected
+one.
+
+## Motivation
+
+Most non-trivial DApps will be distributed not only in the usual sense, but will also have their on-chain logic spread
+across multiple validator scripts. Some types of transactions will then be expected to consume and create UTxOs
+belonging to different scripts.
+
+To take an abstracted example, consider a DApp with only two validators scripts, `S₁` and `S₂`. There are transactions
+`T₁` that deal only with the first validator script, transactions `T₂` that deal only with the second, and
+transactions `T₁₂` that involve both. A `T₁₂` might for example simply move some funds from the first script to the
+second. The question is, how can the `S₁` ensure that the recipient is in fact `S₂`?
+
+The simplest solution is that we first compile `S₂`, then supply its `ValidatorHash` or its `Address` as a parameter
+to `S₁` which can trivially compare it to the recipient's address.
+
+If, however, there is also a valid transaction `T₂₁` that moves funds from `S₂` to `S₁`, we can't apply the same
+solution: there is currently no way to resolve a mutual dependency between scripts.
+
+There are workarounds to guard against the substitution. We can mint a limited amount of a token, for example, supply
+its `CurrencySymbol` as a parameter to all scripts, and then ensure that every UTxO locked by any script in our DApp
+carries one of these tokens as an identification badge. This however introduces unnecessary complexity and increases
+all transaction costs. It would be better to solve the root issue of mutual dependency.
+
+## Specification
```

Comment:
This specification is not specific enough. I can't figure out whether you're just proposing changes to off-chain code or what. It seems like you _must_ want some changes for on-chain code but I can't figure out what, and that's the most important thing.

Reply:
The new field of `TxInfo`, namely 
```
txInfoValidators :: TxInfo -> [ValidatorHash]
```
is the proposed on-chain API modification. How should I make it clearer?

Reply:
What does it do? Can you just explain specifically what is in that field and how it solves the problem?

Reply:
That's in there as well, but I certainly could have made the explanation more prominent:

> The `[ValidatorHash]` list calculated by `allScripts` would be made available as the `txInfoValidators` accessor so each validator triggered by the transation could check it against the addresses of transaction's UTxOs.

In other words, the on-chain `txInfoValidators` field would contain `validatorHashes vals` where `vals` is the argument supplied off-chain to `allScripts` as part of the transaction lookups. The resulting list of validator hashes would become a part of `TxInfo`.

Reply:
I'm sorry, that doesn't make any sense to me. The fields of `TxInfo` must be computed from the transaction itself. How is this field computed? How does the function `mkTxInfo :: Tx -> TxInfo` change?

Reply:
As we build the transaction, the `allScripts vals` lookup will presumably result in a new constructor of `ScriptLookups`, say
```
slValidators :: [Validator]
```
The same value would be carried in `Tx` as a field: 
```
txValidators :: [Validator]
```
To compute the `txInfoValidators` field, we just apply `validatorHashes` to this, as in `TxInfo{..., txInfoValidators= validatorHashes (txValidators tx)}`. Is this clear enough?

Reply:
Okay, so you're adding a new field to transactions. That's a big deal! I suggest you drop the stuff about the Plutus API and focus on that: how are you changing transactions and transaction validation. The rest is important additional information, but not the core change.